PORTING.md

```diff
@@ -0,0 +1,128 @@
+## Envoy Porting
```

Comment:
What is the goal of this doc? To capture some known port issues, to provide best practices for not breaking ports, to advise on how to add a new port?

Reply:
My #1 goal in authoring PORTING.md is to help regular contributors to look beyond baked-in Linux x86_64 assumptions and anticipate what is needed to write portable C++ code. To that end, it needs be structured to help resolve CI pipeline breakage, whether that happens to their submission in the OS/X or Windows ports, or ARM 32 or 64, or Android or other future ports.

Reply:
Should the key takeaways just become part of STYLE.md? I get that we want a more details doc providing an explanation of "why?" rather than just "how", but it would be helpful to only be able to read STYLE.md and have the rules of the road here.

Reply:
I would leave it to the core maintainers to decide which suggestions to adopt. For example, suggesting number of bytes or number of elements (.size() or .count()) used for a storage class in-memory should always be represented as size_t/ptrdiff_t, while sizes of storage class on-disk should be represented as off_t, because that's what ISO stdc++ does. Fighting with the C++ standard library is probably the silliest windmill we could tilt at.

Reply:
@envoyproxy/maintainers for thoughts on this. I think we'd be happy to take some basic suggestions like this. PR welcome, we can bikeshed there.